Questions:
1. What percentage of all licensed sets ever released were Star Wars themed?

2. In which year was Star Wars not the most popular licensed theme (in terms of number of sets releasel that year)?

In [1]:
import polars as pl

In [2]:
df = pl.read_csv('lego_sets.csv')
parent_themes = pl.read_csv('parent_themes.csv')


missing **set_num** indicates a duplicate or invalid row

In [3]:
df = df.filter(pl.col('set_num').is_not_null())

In [4]:
merged = parent_themes.filter(pl.col('is_licensed')).join(df,how='left',left_on='name',right_on='parent_theme')

In [5]:
num_licenced_starwars_themed_legos = merged.filter(pl.col('name').str.contains('Star Wars')).shape[0]

In [6]:
num_all_licencedlegos = merged.shape[0]

In [7]:
answer1 = round(num_licenced_starwars_themed_legos/num_all_licencedlegos * 100,2)
answer1

51.65

The percentage of all licensed sets ever released that were Star Wars themed is **51.65**.

Q2

In [8]:
df_licensed = df.join(parent_themes.filter(pl.col('is_licensed')),left_on='parent_theme',right_on='name',how='inner')

In [10]:
df_licensed.group_by(pl.col(['year','parent_theme'])).len().sort(['year','len']).unique(subset='year',keep='last').sort('year').rename({'len':'most_popular'}).filter(pl.col('parent_theme') != 'Star Wars')

shape: (1, 3)
┌──────┬──────────────┬──────────────┐
│ year ┆ parent_theme ┆ most_popular │
│ ---  ┆ ---          ┆ ---          │
│ i64  ┆ str          ┆ u32          │
╞══════╪══════════════╪══════════════╡
│ 2017 ┆ Super Heroes ┆ 72           │
└──────┴──────────────┴──────────────┘

The most popluar, licensed theme surpassing Star Wars is **Super Heroes** and it happend in **2017** with **72** releases.